publication-graph-maintenance/main.js

```diff
@@ -66,52 +67,64 @@ async function filterActualChangesToPublicationGraph(delta){
   // wrong results, because the insert:s1 won't be considered an effective insert.
   // I.e. we end up with delta: [ deletes: s1 ] to execute against the publication graph.
   const dbEndpoint = SKIP_MU_AUTH_DELTA_FOLDING ? VIRTUOSO_ENDPOINT : MU_AUTH_ENDPOINT;
-  const foldedDelta = await foldChangeSet(delta, { dbEndpoint });
+  const foldedDelta = await foldChangeSet(delta, {dbEndpoint});
   const foldedDeletes = chain(foldedDelta).map(c => c.deletes).flatten().value();
   const foldedInserts = chain(foldedDelta).map(c => c.inserts).flatten().value();
-  const actualDeletes = [];
-
-  //From this folded information, we now check wether the publication graph needs an update
-  for(const triple of foldedDeletes){
-    if((await tripleExists(triple, PUBLICATION_GRAPH, { dbEndpoint })) ){
-      actualDeletes.push(triple);
-      await new Promise(r => setTimeout(r, UPDATE_PUBLICATION_GRAPH_SLEEP)); //performance consideration
-    }
-  }
 
-  const actualInserts = [];
+  const actualDeletes = await getActualDeletes(foldedDeletes, {dbEndpoint});
+  const actualInserts = await getActualInserts(foldedInserts, {dbEndpoint});
 
-  for(const triple of foldedInserts){
-    if( !(await tripleExists(triple, PUBLICATION_GRAPH, { dbEndpoint })) ){
-      actualInserts.push(triple);
-      await new Promise(r => setTimeout(r, UPDATE_PUBLICATION_GRAPH_SLEEP));
-    }
-  }
-
-  if(!(actualInserts.length || actualDeletes.length)){
+  if (!(actualInserts.length || actualDeletes.length)) {
     return [];
-  }
-  else {
-    return [ { deletes: actualDeletes, inserts: actualInserts } ];
+  } else {
+    return [{deletes: actualDeletes, inserts: actualInserts}];
   }
 
 }
 
-async function tripleExists( tripleObject, graph, config ){
-  const tripleStr = serializeTriple(tripleObject);
-  const existsQuery = `
-    ASK {
-      GRAPH ${sparqlEscapeUri(graph)}{
-        ${tripleStr}
-      }
-    }
-  `;
+async function getActualDeletes(deletes, {dbEndpoint}) {
+  return await findMatchesInStore(deletes, {
+    graph: PUBLICATION_GRAPH,
+    dbEndpoint
+  });
+}
 
-  const result = await query(existsQuery, {}, config.dbEndpoint);
-  return result.boolean;
+async function getActualInserts(inserts, {dbEndpoint}) {
+  const inPublicationGraph = await findMatchesInStore(inserts, {
+    graph: PUBLICATION_GRAPH,
+    dbEndpoint
+  });
+  return inserts.filter(t1 => !inPublicationGraph.find(t2 => compareTriple(t1, t2)));
```

Comment:
Given the potential issues with compareTriples, could we think of delegating the detection of real inserts AND inserting them as a query in the database?